llvm/lib/MC/GOFFObjectWriter.cpp

```diff
@@ -239,6 +298,63 @@ class GOFFWriter {
 GOFFWriter::GOFFWriter(raw_pwrite_stream &OS, MCAssembler &Asm)
     : OS(OS), Asm(Asm) {}
 
+void GOFFWriter::defineSectionSymbols(const MCSectionGOFF &Section) {
+  if (Section.isSD()) {
+    GOFFSymbol SD(Section.getName(), Section.getId(),
+                  Section.getSDAttributes());
+    writeSymbol(SD);
+  }
+
+  if (Section.isED()) {
+    GOFFSymbol ED(Section.getName(), Section.getId(),
+                  Section.getParent()->getId(), Section.getEDAttributes());
+    if (Section.requiresLength())
+      ED.SectionLength = Asm.getSectionAddressSize(Section);
+    writeSymbol(ED);
+  }
+
+  if (Section.isPR()) {
+    GOFFSymbol PR(Section.getName(), Section.getId(),
+                  Section.getParent()->getId(), Section.getPRAttributes());
+    PR.SectionLength = Asm.getSectionAddressSize(Section);
+    if (Section.requiresNonZeroLength()) {
```

Comment:
This is also unclear to me.   What are the actual rules?  Is is that PR sections of size zero are generally not allowed?  Then why do we need an extra flag here?

Reply:
I see.  I think we should implement this rule then - if a section is the target of an ADA link, it cannot be empty.  That would be preferable to upper layers having to remember setting this flag.  (I guess we could still have a flag internally, but that would get automatically set e.g. by `setADA`.)

Or, in the alternative, should we simply *not* set the ADA link if the target section is empty?  If an ADA section, it cannot actually be used for anything, so do we even need to register it as ADA for a text section?

Reply:
> Or, in the alternative, should we simply not set the ADA link if the target section is empty? If an ADA section, it cannot actually be used for anything, so do we even need to register it as ADA for a text section?

That is a simple solution, too. I am not sure if this works with the HLASM output.

Reply:
> That is a simple solution, too. I am not sure if this works with the HLASM output.

Ok, I contradict myself. Setting the ADA to null gets around the binder error but other parts assume that there is always an ADA. E.g. when calling an external functions.

Reply:
But if you do call an external function, don't you then have to have actual data in the ADA identifying that external function?   I.e. don't those places that require a non-null ADA also automatically imply there are contents?

Reply:
For this specific case I do not know what happens under the hood.
For an externally callable function, a function descriptor is generated in the ADA by the _binder_. From the LLVM point of view, the ADA would still be empty. The simple example is:
```
void empty() {}
```
Here, the ADA is empty. Still, the binder generates the function descriptor. A caller of this function can use a function descriptor in its own compilation unit:
```
	DC	R(empty)
	DC	V(empty)
```
I would expect that I get a binder error on `R(empty)` if no associated data area was given for the code section. But I have not tried.

Reply:
I talked to a colleague. Exactly this ^^^ case will not work. The binder gives no warning/error but it will cause problems at runtime because the environment parameter is zero.

Reply:
OK, thanks for following up on this.